pkg/apis/gcp/database/v1alpha1/cloudsql_instance_types.go

```diff
@@ -164,33 +176,97 @@ func NewCloudSQLInstanceSpec(properties map[string]string) *CloudsqlInstanceSpec
+	spec.DatabaseVersion = properties["databaseVersion"]
+	spec.Labels = util.ParseMap(properties["labels"])
+	spec.Region = properties["region"]
+	spec.StorageType = properties["storageType"]
+	if v, err := strconv.Atoi(properties["storageGB"]); err != nil {
+		spec.StorageGB = DefaultStorageGB
+	} else {
+		spec.StorageGB = int64(v)
+	}
+	spec.Tier = properties["tier"]
+	return spec
+}
 
-	val, ok := properties["tier"]
-	if ok {
-		spec.Tier = val
+// ConnectionSecret returns a connection secret for this instance
+func (i *CloudsqlInstance) ConnectionSecret() *corev1.Secret {
+	s := resource.ConnectionSecretFor(i, CloudsqlInstanceGroupVersionKind)
+	s.Data[corev1alpha1.ResourceCredentialsSecretEndpointKey] = []byte(i.Status.Endpoint)
+	s.Data[corev1alpha1.ResourceCredentialsSecretUserKey] = []byte(i.DatabaseUserName())
+	return s
+}
+
+// DatabaseInstance representing spec of this instance
+func (i *CloudsqlInstance) DatabaseInstance(name string) *sqladmin.DatabaseInstance {
+	authnets := make([]*sqladmin.AclEntry, len(i.Spec.AuthorizedNetworks))
+	for i, v := range i.Spec.AuthorizedNetworks {
+		authnets[i] = &sqladmin.AclEntry{Value: v}
 	}
 
-	val, ok = properties["region"]
-	if ok {
-		spec.Region = val
+	return &sqladmin.DatabaseInstance{
+		Name:            name,
+		Region:          i.Spec.Region,
+		DatabaseVersion: i.Spec.DatabaseVersion,
+		Settings: &sqladmin.Settings{
+			Tier:           i.Spec.Tier,
+			DataDiskType:   i.Spec.StorageType,
+			DataDiskSizeGb: i.Spec.StorageGB,
+			IpConfiguration: &sqladmin.IpConfiguration{
+				AuthorizedNetworks: authnets,
+			},
+			UserLabels: i.Spec.Labels,
+		},
 	}
+}
 
-	val, ok = properties["databaseVersion"]
-	if ok {
-		spec.DatabaseVersion = val
+// DatabaseUserName returns default database user name base on database version
+func (i *CloudsqlInstance) DatabaseUserName() string {
+	if strings.HasPrefix(i.Spec.DatabaseVersion, PostgresqlDBVersionPrefix) {
+		return PostgresqlDefaultUser
 	}
+	return MysqlDefaultUser
+}
+
+// GetResourceName based on the NameFormat spec value,
+// If name format is not provided, resource name defaults to UID
+// If name format provided with '%s' value, resource name will result in formatted string + UID,
+//   NOTE: only single %s substitution is supported
+// If name format does not contain '%s' substitution, i.e. a constant string, the
+// constant string value is returned back
+//
+// Examples:
+//   For all examples assume "UID" = "test-uid"
+//   1. NameFormat = "", ResourceName = "test-uid"
+//   2. NameFormat = "%s", ResourceName = "test-uid"
+//   3. NameFormat = "foo", ResourceName = "foo"
+//   4. NameFormat = "foo-%s", ResourceName = "foo-test-uid"
+//   5. NameFormat = "foo-%s-bar-%s", BucketName = "foo-test-uid-bar-%!s(MISSING)"
+func (i *CloudsqlInstance) GetResourceName() string {
+	return util.ConditionalStringFormat(i.Spec.NameFormat, string(i.GetUID()))
+}
+
+// IsAvailable for usage/binding
```

Comment:
Is this function used anywhere? In #545 I updated the resource claim controller to consider resources ready for binding when their binding phase is `unbound`, rather than when `IsAvailable` is true.

Reply:
Yes, it is. I see what the possible confusion is.